third_party/tlx/tutorials/gemm-WS-hopper.py

```diff
@@ -0,0 +1,247 @@
+    offset_bn = pid_n * BN
+
+    # Need NUM_STAGES sets of SMEM buffers for A and B
+    # where each set contains two for A and one for B.
+    # Split A into two in M-dimension to have two consumer tasks for wgmma
+    a = tlx.local_alloc((BM // 2, BK), tl.float16, NUM_STAGES * 2)
+    b = tlx.local_alloc((BK, BN), tl.float16, NUM_STAGES)
+
+    # Need NUM_STAGES sets of mbarriers for A and B
+    # where each set contains two for A and one for B.
+    # Do the above for both empty states and full states respectively.
+    bars_empty_a = tlx.alloc_barriers(num_barriers=NUM_STAGES * 2,)
+    bars_full_a = tlx.alloc_barriers(num_barriers=NUM_STAGES * 2, arrive_count=1)
+    bars_empty_b = tlx.alloc_barriers(num_barriers=NUM_STAGES, arrive_count=2)
+    bars_full_b = tlx.alloc_barriers(num_barriers=NUM_STAGES, arrive_count=1)
+
+    # Warp specilization
+    with tlx.async_tasks():
+        # Producer (async load)
+        with tlx.async_task("default"):
+            # Assuming NUM_STAGES = 2
+            # p should be 1, 1, 0, 0, 1, 1, 0, 0, ...
+            p = 1
+
+            for k in range(0, tl.cdiv(K, BK)):
+                buf = k % NUM_STAGES
+                offset_k = k * BK
+
+                # Async load to a[buf]
+                empty_a_1st = tlx.local_view(bars_empty_a, buf)  # mbar
+                full_a_1st = tlx.local_view(bars_full_a, buf)  # mbar
+                tlx.barrier_wait(bar=empty_a_1st, phase=p)  # EmptyBar A1 wait
+                tlx.barrier_expect_bytes(full_a_1st, (BM//2) * BK * 2)
+                data_a_1st = tlx.local_view(a, buf)  # smem data
+                tlx.async_descriptor_load(
+                    desc_in_1,
+                    data_a_1st,
+                    [offset_am, offset_k],
+                    full_a_1st)
+
+                # Async load to b[buf]
+                empty_b = tlx.local_view(bars_empty_b, buf)
+                full_b = tlx.local_view(bars_full_b, buf)
+                tlx.barrier_wait(bar=empty_b, phase=p)
+                tlx.barrier_expect_bytes(full_b, BN * BK * 2)
+                data_b = tlx.local_view(b, buf)
+                tlx.async_descriptor_load(
+                    desc_in_2,
+                    data_b,
+                    [offset_k, offset_bn],
+                    full_b)
+
+                # Async load to a[buf+NUM_STAGES]
+                empty_a_2nd = tlx.local_view(bars_empty_a, buf+NUM_STAGES)
+                full_a_2nd = tlx.local_view(bars_full_a, buf+NUM_STAGES)
+                tlx.barrier_wait(bar=empty_a_2nd, phase=p)
+                tlx.barrier_expect_bytes(bar=full_a_2nd, size=(BM//2) * BK * 2)
+                data_a_2nd = tlx.local_view(a, buf+NUM_STAGES)  # smem data
+                tlx.async_descriptor_load(
+                    desc_in_1,
+                    data_a_2nd,
+                    [offset_am + (BM//2), offset_k],
+                    full_a_2nd)
+
+                # Flip phase after every NUM_STAGES iterations finish
```

Comment:
This likely results in conditionals. I worked it around with the following logics in https://github.com/facebookexperimental/triton/pull/205

          acc_cnt = 0
            for _ in tl.range(lo, hi, BLOCK_N):
                buf_id = acc_cnt % NUM_BUFFERS
                # buffers in a row share the same phase
                p = p ^ (buf_id == 0)
                ...
               acc_cnt +=1